host/crypto/bcrypt/cert.c

```diff
@@ -0,0 +1,1295 @@
+    key_handle = NULL;
+
+    result = OE_OK;
+
+done:
+    if (key_handle)
+        BCryptDestroyKey(key_handle);
+
+    if (key_alg_name)
+        free(key_alg_name);
+
+    return result;
+}
+
+oe_result_t oe_cert_get_ec_public_key(
+    const oe_cert_t* cert,
+    oe_ec_public_key_t* public_key)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    BCRYPT_KEY_HANDLE key_handle = NULL;
+    PWSTR key_alg_name = NULL;
+    ULONG key_alg_name_size;
+
+    /* Clear public key for all error pathways */
+    if (public_key)
+        oe_secure_zero_fill(public_key, sizeof(oe_ec_public_key_t));
+
+    /* Reject invalid parameters */
+    if (!cert || !public_key)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    OE_CHECK(_bcrypt_get_public_key_from_cert(cert, &key_handle));
+
+    /* Check if EC key */
+    result = _bcrypt_get_key_algorithm(
+        key_handle, &key_alg_name, &key_alg_name_size);
+    if (result != OE_OK)
+        OE_RAISE(OE_PUBLIC_KEY_NOT_FOUND);
+
+    if (wcsncmp(
+            key_alg_name,
+            BCRYPT_ECDSA_ALGORITHM,
+            _countof(BCRYPT_ECDSA_ALGORITHM) - 1) != 0 &&
+        wcsncmp(
+            key_alg_name,
+            BCRYPT_ECDH_ALGORITHM,
+            _countof(BCRYPT_ECDH_ALGORITHM) - 1) != 0)
+    {
+        OE_RAISE(OE_PUBLIC_KEY_NOT_FOUND);
+    }
+
+    /* Initialize the EC public key */
+    oe_ec_public_key_init(public_key, key_handle);
+    key_handle = NULL;
+
+    result = OE_OK;
+
+done:
+    if (key_handle)
+        BCryptDestroyKey(key_handle);
+
+    if (key_alg_name)
+        free(key_alg_name);
+
+    return result;
+}
+
+/* Used by tests/crypto/ec_tests|rsa_tests */
+oe_result_t oe_cert_chain_get_length(
+    const oe_cert_chain_t* chain,
+    size_t* length)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    const cert_chain_t* impl = (const cert_chain_t*)chain;
+
+    /* Clear the length (for failed return case) */
+    if (length)
+        *length = 0;
+
+    /* Reject invalid parameters */
+    if (!_cert_chain_is_valid(impl) || !length)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Get the number of certificates in the chain */
+    if (impl->cert_chain->cChain == 0 || !impl->cert_chain->rgpChain ||
+        impl->cert_chain->rgpChain[0]->cElement == 0)
+        OE_RAISE_MSG(
+            OE_CRYPTO_ERROR, "No certs found in oe_cert_chain_t impl\n", NULL);
+
+    *length = (size_t)impl->cert_chain->rgpChain[0]->cElement;
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_cert_chain_get_cert(
+    const oe_cert_chain_t* chain,
+    size_t index,
+    oe_cert_t* cert)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    const cert_chain_t* impl = (const cert_chain_t*)chain;
+    size_t length;
+    PCCERT_CONTEXT found_cert = NULL;
+
+    /* Clear the output certificate for all error pathways */
+    if (cert)
+        memset(cert, 0, sizeof(oe_cert_t));
+
+    /* Reject invalid parameters */
```

Comment:
No, this is an intentional pattern that Mike introduced where the expectation for functions are that return parameters are zeroed even on failure cases, so the clears happen even before the invalid param checks. Comment is outdated in this case since this has been combined into a single switch case in #2072 already.

Reply:
Ok